src/main/java/uk/gov/justice/digital/hmpps/whereabouts/services/court/VideoLinkMigrationService.kt

```diff
@@ -15,49 +17,106 @@ import uk.gov.justice.digital.hmpps.whereabouts.model.VideoLinkBookingEventType
 import uk.gov.justice.digital.hmpps.whereabouts.repository.VideoLinkAppointmentRepository
 import uk.gov.justice.digital.hmpps.whereabouts.repository.VideoLinkBookingEventRepository
 import uk.gov.justice.digital.hmpps.whereabouts.repository.VideoLinkBookingRepository
+import uk.gov.justice.digital.hmpps.whereabouts.services.court.events.OutboundEvent
+import uk.gov.justice.digital.hmpps.whereabouts.services.court.events.OutboundEventsService
 import java.time.LocalDate
 import java.time.LocalDateTime
 import java.time.LocalTime
+import java.util.concurrent.CompletableFuture
+import kotlin.system.measureTimeMillis
 
 @Service
+@Transactional(readOnly = true)
 class VideoLinkMigrationService(
   private val videoLinkBookingRepository: VideoLinkBookingRepository,
   private val videoLinkBookingEventRepository: VideoLinkBookingEventRepository,
   private val videoLinkAppointmentRepository: VideoLinkAppointmentRepository,
+  private val outboundEventsService: OutboundEventsService,
 ) {
   companion object {
     private val log: Logger = LoggerFactory.getLogger(this::class.java)
   }
 
-  @Transactional(readOnly = true)
+  /**
+   * This service is only used in the migration process for video link bookings and not used in any user flow.
+   * It can be removed along with other video link booking features when that migration has been completed.
+   */
+  @Async("asyncExecutor")
+  fun migrateVideoLinkBookingsSinceDate(
+    fromDate: LocalDate,
+    pageSize: Int,
+    waitMillis: Long,
+  ): CompletableFuture<MigrationSummary> {
+    val task = CompletableFuture<MigrationSummary>()
+    var numberOfEventsRaised = 0
+    val pageable = Pageable.ofSize(pageSize)
+
+    log.info("Starting migration of video link bookings from $fromDate at ${LocalDateTime.now()}")
+
+    val elapsed = measureTimeMillis {
+      var page = videoLinkBookingEventRepository.findAllByMainStartTimeGreaterThanAndEventTypeEquals(
+        LocalDateTime.of(fromDate.year, fromDate.month, fromDate.dayOfMonth, 0, 0),
+        VideoLinkBookingEventType.CREATE,
+        pageable.first(),
+      )
+
+      val numberOfBookings = page.totalElements
+      val numberOfPages = page.totalPages
+
+      while (page.hasContent()) {
+        page.content.map {
+          outboundEventsService.send(OutboundEvent.VIDEO_LINK_BOOKING_MIGRATE, it.videoLinkBookingId)
+          numberOfEventsRaised++
+        }
+
+        if (page.isLast) {
+          break
+        }
+
+        if (waitMillis > 0) {
+          Thread.sleep(waitMillis)
+        }
+
+        page = videoLinkBookingEventRepository.findAllByMainStartTimeGreaterThanAndEventTypeEquals(
+          LocalDateTime.of(fromDate.year, fromDate.month, fromDate.dayOfMonth, 0, 0),
+          VideoLinkBookingEventType.CREATE,
+          pageable.next(),
+        )
+      }
+
+      task.complete(MigrationSummary(numberOfBookings, numberOfPages, numberOfEventsRaised))
```

Comment:
Won't it be finished by the time it gets here anyway? i.e is the completable future necessary?

Reply:
Async endpoints return a completableFuture<type>, but nothing is actually waiting for it.

Reply:
Agreed, it would if it returned immediately but this isn't is all. By the time it gets here it's finished.

Reply:
I seem to remember trying with a void return type and it wasn't happy, so need the completable future, but I'll re-check.

Reply:
Async needs a completableFuture - but this can be Void or Another class. It could be made to be void, but it is useful to test against.